Release/1.49 blogpost (#217)
conan-io/conan-io.github.io: _posts/2022-06-16-New-conan-release-1-49.markdown

---

<script type="application/ld+json">
{ "@context": "https://schema.org", 
 "@type": "TechArticle",
 "headline": "Version 1.49 of Conan C++ Package Manager is Released",
 "alternativeHeadline": "Learn all about the new 1.49 Conan C/C++ package manager version",
 "image": "https://docs.conan.io/en/latest/_images/frogarian.png",
 "author": "Conan Team", 
 "genre": "C/C++", 
 "keywords": "c c++ package manager conan release", 
 "publisher": {
    "@type": "Organization",
    "name": "Conan.io",
    "logo": {
      "@type": "ImageObject",
      "url": "https://media.jfrog.com/wp-content/uploads/2017/07/20134853/conan-logo-text.svg"
    }
},
 "datePublished": "2022-06-16",
 "description": "Removed Python2 support, improved CMakePresets, new [layout] section in conanfile.txt, new tools.apple.fix_apple_shared_install_name tool, new "can_run()", "check_min_vs" helpers.",
 }
</script>

We are pleased to announce that [Conan 1.49 is
out](https://github.com/conan-io/conan/releases/tag/1.49.0) and comes with some
significant new features and bug fixes. First, it's worth mentioning that we finally

> removed support for Python2 in Conan. Also, we made some improvements to the CMakePresets

Reviewer: ```suggestion
removed support for Python2 in Conan. Also, we have made some improvements to the CMakePresets
```

support to allow parameterizing the output folders using the value of the recipe settings
and options. Now you can use layouts feature in the *conanfile.txt* adding the `[layout]`

> section. We also added a new `tools.apple.fix_apple_shared_install_name` tool, to fix shared

Reviewer: > 
```suggestion
section. A new `tools.apple.fix_apple_shared_install_name` tool has been introduced, to fix shared
```
(just to change the "We have.." style but nothing wrong otherwise!)

libraries that do not set the correct `LC_ID_DYLIB` and `LC_LOAD_DYLIB` fields. Finally,

> we added two new helpers: `conan.tools.build.can_run()` and

Reviewer: ```suggestion
we have added two new helpers: `conan.tools.build.can_run()` and
```

> `conan.tools.microsoft.check_min_vs`.

Reviewer: ```suggestion
`conan.tools.microsoft.check_min_vs()`.
```

## Removed Python2 support

> From version 1.49, Conan will not work with Python 2. This is because [security

Reviewer: ```suggestion
Starting with version 1.49, Conan will no longer work with Python 2. This is because [security
```

vulnerabilities of Conan dependencies](https://github.com/advisories/GHSA-ffqj-6fqr-9h24)
that haven't been addressed in Python 2, so the only alternative moving forward is to
finally remove Python 2 suport.

> Python 2 was officially declared End Of Life 2 years and a half now, and [Conan

Reviewer: ```suggestion
Python 2 was officially declared End Of Life two years and a half ago, and [Conan
```

1.22](https://docs.conan.io/en/latest/changelog.html#id1006) already declared Python 2 as
not supported. Extra blockers have been added in previous Conan releases to make everyone

> aware. Now the security vulnerabilities that are out of our scope, makes impossible to

Reviewer: ```suggestion
aware. With the security vulnerabilities that fall outside of our scope, it has become impossible to
```

move forward support for Python 2. 

**Please upgrade to Python>=3.6 to continue using Conan>=1.49.** 

## New [layout] section in conanfile.txt

Until Conan 1.49 you could only use the layouts

> [feature](https://docs.conan.io/en/latest/reference/conanfile/tools/layout.html) defining

Reviewer: ```suggestion
[feature](https://docs.conan.io/en/latest/reference/conanfile/tools/layout.html) by defining
```

default `cmake_layout` and the *CMakeUserPresets.json* is created in the base folder,
next to the *CMakeLists.txt*.

```txt
├── CMakeLists.txt
├── CMakeUserPresets.json
├── build
│   └── generators
│       ├── CMakePresets.json
│       ├── FindZLIB.cmake
│       ├── ...
│       └── module-ZLIBTargets.cmake
├── conanfile.txt
└── src
    └── compressor.c
```

Then you can invoke the presets calling CMake:

```txt
cmake --preset release
cmake --build --preset release      
```

The results of the build will end in the same folder as if we were using a *conanfile.py*
with the `cmake_layout` defined in the `layout()` method:

```txt
.
├── CMakeLists.txt
├── CMakeUserPresets.json
├── build
│   ├── Release
│   │   ├── CMakeCache.txt
│   │   ├── CMakeFiles
│   │   │   └── ...
│   │   ├── Makefile
│   │   ├── cmake_install.cmake
│   │   └── compressor
│   └── generators
│       ├── CMakePresets.json
│       ├── ...
│       └── module-ZLIBTargets.cmake
├── conanfile.txt
└── src
    └── compressor.c
```

You can check that the *compressor* application was built in the `/build/Release` folder as
defined in the `cmake_layout`.

## Improved CMakePresets with parameterized layout

> Conan has incrementally improved the support for CMakePresets during the last releases.

Reviewer: ```suggestion
Conan has incrementally improved the support for CMakePresets in the most recent releases.
```

default values for the `conanfile.folders.build` and `conanfile.folders.generators`
attributes and also the default preset names for CMakePresets. Let's begin from the
previous `[layout]` example to see how this works.

As you know, the name of the presets and output folders by default is set to the value of
the `build_type` setting, so they are named `Release` or `Debug` by default. Let's use
this configuration to change that default, so that those folders and preset names take
other settings into account, like for example the `arch` setting. 

Let's to do a `conan install` and pass the `tools.cmake.cmake_layout:build_folder_vars`
configuration as an argument:

```bash
conan install . -c tools.cmake.cmake_layout:build_folder_vars='["settings.arch", "settings.build_type"]'
```

Now, if you list the available presets, you will see a new preset including the
architecture and the builds will output the files to the corresponding folder containing
the used archicture and build type in the name.

```bash
$ cmake --list-presets
Available configure presets:

  "x86_64-release-release" - 'x86_64-release-release' config
  "release"                - 'release' config

> $ cmake --preset x86_64-release-release

Reviewer 2: Is this correct with 2 ``release-release``? reads a bit weird. Maybe we want to have another example with ``compiler``  or ``shared`` instead?

Author: My mistake, I was running this in develop2 branch, looks like the problem of the names only happens in develop2 branch but not in 1.X. 

cc/ @lasote

Reviewer 2: Maybe an issue in the develop->develop2 merge?

Reviewer 3: I'll take a look today at some point.

...

$ cmake --build --preset Release-x86_64
...
```

## New tools.apple.fix_apple_shared_install_name tool

We have added the new tool
[tools.apple.fix_apple_shared_install_name](https://docs.conan.io/en/latest/reference/conanfile/tools/apple.html#conan-tools-apple-fix-apple-shared-install-name)
to help users address problems that can occur when shared libraries in Apple platforms.

As we explained in the [Conan documentation](https://docs.conan.io/en/latest/reference/conanfile/tools/gnu/autotools.html#a-note-about-relocatable-shared-libraries-in-macos-built-the-autotools-build-helper)
making shared libraries relocatable in Apple platforms requires an extra effort because

> when these kinds of libraries are built, two sections called `LC_ID_DYLIB` and

Reviewer: ```suggestion
when shared libraries are built, two fields called `LC_ID_DYLIB` and
```

> `LC_LOAD_DYLIB` are added to the information embedded in the library usually pointing to

Reviewer: ```suggestion
`LC_LOAD_DYLIB`  containing the "install name" are embedded in the library header information, usually pointing to
```

an absolute shared library location in the filesystem. This absolute location will make
the library load fail when it's loaded in a different system without the library in that
location. 

The best solution to avoid this problem is to pass the appropriate flags for the linker so
that the absolute folder is substituted with the `@rpath` value. The `@rpath` special
keyword will tell the loader to search a list of paths to find the library. These paths
can be defined by the consumer of that library by defining the `LC_RPATH` field through

> linker options. 

Reviewer: (Not sure it's worth mentioning that CMake will typically do this for users for binaries in the build tree - that is, CMake analyses the location in disk of the libraries you link against, and adds LC_RPATH entries pointing to those locations such that you can run the binaries without relying on DYLD_LIBRARY_PATH or similar)

Author: Yes, good point, I'll add a comment about that

Not all libraries are prepared to provide relocatable shared libraries, so if you make a
package for a library whose binaries are not relocatable, you have only two options,
either patching the libraries' build scripts to make them relocatable or fixing the
binaries after the build. 

> The `tools.apple.fix_apple_shared_install_name` tool can help to fix the binaries after

Reviewer: ```suggestion
The `tools.apple.fix_apple_shared_install_name` tool can help fix the binaries after
```
